synapse/events/spamcheck.py

```diff
@@ -270,7 +275,7 @@ def register_callbacks(
 
     async def check_event_for_spam(
         self, event: "synapse.events.EventBase"
-    ) -> Union[Decision, str]:
+    ) -> Union[Decision, Tuple[Codes, Dict], str]:
```

Comment:
This seems like a super confusing api. why are there four different potential return types? (`Decision` is itself a Union of two types)

It also doesn't seem to correspond to the docstring.

Reply:
(it's this function's job to flatten the myriad different potential return types from the module callback into a sane api for the rest of synapse. It seems to be spectacularly failing at that job currently)